### PR TITLE
[QNN EP] Fix Batch Normalization Op Builder 

### PR DESCRIPTION
**Pull request by weischan-quic**
### Description
There is a gap between onnx’s definition of batch normalization and QNN’s.

According to the formula:
onnx: `(X - input_mean) / sqrt(input_var + epsilon) * scale + B`
QNN: `X * weight + bias`

We can then deduce that:
`weight = scale / sqrt(var + epsilon)`
`bias = B – (mean * scale / sqrt(var + epsilon))`

We must calculate the weight and bias, and their quantization parameters for QNN in QNN EP.
Therefore, `scale`, `B`, `input_mean`, and `input_var` must be static (`initializer`).

Implementation:
Firstly, dequantize `scale`, `B`, `input_mean`, and `input_var` to floating point.
Second, calculate `weight` and `bias`, and their quantization parameters.
Finally, quantize `weight` and `bias`, and add them into `TensorWrapper`

### Motivation and Context
Fix QnnHTPBackendTests.BatchNorm1D and QnnHTPBackendTests.BatchNorm2D failures

### CODE REVIEW

**Comment by weischan-quic** [comment]
@microsoft-github-policy-service agree company="Qualcomm"

**Comment by jywu-msft** [comment]
Thanks for the fix!

**Comment by HectorSVC** [comment]
Should re-enable the disabled tests for BatchNorm by removing "DISABLED_" from the test names:
TEST_F(QnnHTPBackendTests, DISABLED_BatchNorm1D)
TEST_F(QnnHTPBackendTests, DISABLED_BatchNorm2D)
https://github.com/microsoft/onnxruntime/blob/a2c62832748184405ab4ddd341a166ffb1f2fafa/onnxruntime/test/providers/qnn/batch_norm_htp_test.cc#L184

**Comment by weischan-quic** [comment]
> Should re-enable the disabled tests for BatchNorm by removing "DISABLED_" from the test names: TEST_F(QnnHTPBackendTests, DISABLED_BatchNorm1D) TEST_F(QnnHTPBackendTests, DISABLED_BatchNorm2D)
> 
> https://github.com/microsoft/onnxruntime/blob/a2c62832748184405ab4ddd341a166ffb1f2fafa/onnxruntime/test/providers/qnn/batch_norm_htp_test.cc#L184

Sure, I will do that.

**Comment by weischan-quic** [comment]
cc @chiwwang

**Comment by weischan-quic** [comment]
Hi @HectorSVC,
I saw CI failures on Windows_CI / Onnxruntime-TVM and Linux_CI / Onnxruntime-TVM.
And the failed test case is `FunctionTest.TestInlinedLocalFunctionNotRemoved`.
Could you please check this failure for us?
And I would like to know how to fix it.
Thanks!

**Comment by jywu-msft** [comment]
> Hi @HectorSVC, I saw CI failures on Windows_CI / Onnxruntime-TVM and Linux_CI / Onnxruntime-TVM. And the failed test case is `FunctionTest.TestInlinedLocalFunctionNotRemoved`. Could you please check this failure for us? And I would like to know how to fix it. Thanks!

the *TVM* CI's are known issues and can be ignored. (that's why they are not marked as required)

**Comment by weischan-quic** [comment]
Thanks for the explanation!

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 1]
```diff
@@ -1,26 +1,488 @@
...
+                       const NodeUnit& node_unit,
+                       const logging::Logger& logger,
+                       std::vector<std::string>& input_names,
+                       bool do_op_validation) const override ORT_MUST_USE_RESULT;
+
   Status IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
                        const NodeUnit& node_unit,
                        const logging::Logger& logger) const override final ORT_MUST_USE_RESULT;
+
+  Status CheckMinMax(float& rmin, float& rmax) const {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:35]: Doesn't look like this function needs to return a `Status`.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 2]
```diff
@@ -1,26 +1,488 @@
...
+    } else if (qnn_data_type == QNN_DATATYPE_UFIXED_POINT_8) {
+      qmin = static_cast<float>(std::numeric_limits<uint8_t>::min());
+      qmax = static_cast<float>(std::numeric_limits<uint8_t>::max());
+    } else {
+      ORT_RETURN_IF(true, "Qnn Data Type: %d not supported yet.", qnn_data_type);
+    }
+    return Status::OK();
+  }
+
+  Status GetQminQmax(const Qnn_DataType_t qnn_data_type,
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:62]: Looks like this function is nearly identical to `GetQminQmax` on line 46. The only difference is the type of the `qmin` and `qmax` output parameters. (To reduce repetition) Perhaps it would be good to use a templated type.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 3]
```diff
@@ -1,26 +1,488 @@
...
+  }
+
+  inline Status GetValueOnQnnDataType(const Qnn_DataType_t qnn_data_type,
+                                      const uint8_t* raw_ptr,
+                                      double& value,
+                                      int& offset) const {
+    switch (qnn_data_type) {
+      case QNN_DATATYPE_INT_8:
+      case QNN_DATATYPE_SFIXED_POINT_8: {
+        value = static_cast<double>(*const_cast<int8_t*>(reinterpret_cast<const int8_t*>(raw_ptr)));
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:103]: Doesn't seem like the `const_cast` should be necessary. I believe the following should also work:
```C++
value = static_cast<double>(*reinterpret_cast<const int8_t*>(raw_ptr));
```

File: onnxruntime/test/providers/qnn/batch_norm_htp_test.cc [hunk 4]
```diff
@@ -181,7 +181,7 @@ static void RunBatchNormQDQTest(const TestInputDef<float>& input_def,
 // Expected val: 1.7755576372146606
 // QNN QDQ val: 2.9963212013244629 (err 1.2207635641098022)
 // CPU QDQ val: 0.82064849138259888 (err 0.95490914583206177)
```
Comment (HectorSVC) [inline onnxruntime/test/providers/qnn/batch_norm_htp_test.cc:183]: Please remove these comments which used to track failure details.

File: onnxruntime/test/providers/qnn/batch_norm_htp_test.cc [hunk 5]
```diff
@@ -199,7 +199,7 @@ TEST_F(QnnHTPBackendTests, DISABLED_BatchNorm1D) {
 // Expected val: 2.8554618358612061
 // QNN QDQ val: 5.3294687271118164 (err 2.4740068912506104)
 // CPU QDQ val: 1.6611330509185791 (err 1.194328784942627)
```
Comment (HectorSVC) [inline onnxruntime/test/providers/qnn/batch_norm_htp_test.cc:201]: remove

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 6]
```diff
@@ -1,26 +1,488 @@
...
+    rmin = std::min(rmin, 0.0f);
+    rmax = std::max(rmax, 0.0f);
+
+    return Status::OK();
+  }
+
+  Status GetQminQmax(const Qnn_DataType_t qnn_data_type,
+                     float& qmin,
+                     float& qmax) const {
+    // onnx only supports 8 bits quantization
```
Comment (HectorSVC) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:49]: We also support 16 bit via Contrib op for now.

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 7]
```diff
@@ -1,26 +1,488 @@
...
+      case QNN_DATATYPE_BOOL_8:
+      case QNN_DATATYPE_STRING:
+      case QNN_DATATYPE_FLOAT_16:
+      default:
+        ORT_RETURN_IF(true, "Qnn Data Type: %d not supported yet.", qnn_data_type);
+    }
+    return Status::OK();
+  }
+
+  inline double dequantize(const OnnxInputInfo& info,
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:316]: nit: ORT convention is to start function names with a capital: `Dequantize`

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 8]
```diff
@@ -1,26 +1,488 @@
...
+                           const double quant_value) const {
+    auto offset = static_cast<double>(info.quant_param.scaleOffsetEncoding.offset);
+    auto scale = static_cast<double>(info.quant_param.scaleOffsetEncoding.scale);
+    return (quant_value + offset) * scale;
+  }
+
+  inline int saturate(const int qmax,
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:323]: nit: ORT convention is to start function names with a capital: `Saturate`

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 9]
```diff
@@ -1,26 +1,488 @@
...
+    if (quant_value > qmax) {
+      return qmax;
+    } else if (quant_value < qmin) {
+      return qmin;
+    } else {
+      return quant_value;
+    }
+  }
+
+  inline float saturate(const float qmax,
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:335]: This function is nearly identical to the one above. Perhaps let's use a templatized type to reduce repetition?

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 10]
```diff
@@ -1,26 +1,488 @@
...
+    if (quant_value > qmax) {
+      return qmax;
+    } else if (quant_value < qmin) {
+      return qmin;
+    } else {
+      return quant_value;
+    }
+  }
+
+  inline Status quantize(const double double_value,
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:347]: nit: Same here `Quantize`

File: onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc [hunk 11]
```diff
@@ -56,32 +513,183 @@ Status BatchNormOpBuilder::IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
...
+      if (input0_info.is_initializer) {
+        ORT_RETURN_IF_ERROR(qnn_model_wrapper.UnpackInitializerData(*input0_info.initializer_tensor, unpacked_tensor));
+      }
+
+      Qnn_TensorType_t tensor_type = GetInputTensorType(qnn_model_wrapper, input0_name);
+      QnnTensorWrapper input_tensorwrapper(input0_name, tensor_type, input0_info.qnn_data_type, input0_info.quant_param,
+                                           std::move(input0_info.shape), std::move(unpacked_tensor));
+      ORT_RETURN_IF_NOT(qnn_model_wrapper.AddTensorWrapper(std::move(input_tensorwrapper)), "Failed to add tensor.");
+    }
+    input_names.push_back(input0_name);
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/opbuilder/batch_norm_op_builder.cc:577]: It looks like this is using the default handling for inputs. If this is the case, we can replace this block with a single call to:

```C++
ORT_RETURN_IF_ERROR(ProcessInput(qnn_model_wrapper, inputs[0], logger, input_names));
```

**Review by adrianlizarraga — COMMENTED** [review]
Thank you for fixing this! A few comments

**Comment by weischan-quic** [comment]
Thanks for the review!
I've update the PR per comments 😄

**Review by adrianlizarraga — APPROVED** [review]

**Review by HectorSVC — APPROVED** [review]
:shipit:
